network/src/network.rs

```diff
@@ -315,39 +297,59 @@ impl Network {
         peers_registry.get(peer_id).map(|peer| {
             let protocol_version = peer.protocol_version(protocol_id);
             SessionInfo {
-                peer: PeerInfo {
-                    peer_id: peer_id.to_owned(),
-                    session_type: peer.session.session_type,
-                    last_ping_time: peer.last_ping_time,
-                    connected_addr: peer.connected_addr.clone(),
-                    identify_info: peer.identify_info.clone(),
-                },
+                peer: peer.clone(),
                 protocol_version,
             }
         })
     }
 
-    pub fn dial_addr(&self, addr: Multiaddr) {
-        if let Err(err) = self.p2p_control.write().dial(addr, DialProtocol::All) {
-            error!(target: "network", "failed to dial: {}", err);
-        }
-    }
-
-    pub fn dial(&self, expected_peer_id: &PeerId, mut addr: Multiaddr) {
+    fn inner_dial(
+        &self,
+        expected_peer_id: &PeerId,
+        mut addr: Multiaddr,
+        dial_protocol: DialProtocol,
+    ) -> Result<(), Error> {
         if expected_peer_id == self.local_peer_id() {
             debug!(target: "network", "ignore dial to self");
-            return;
+            return Ok(());
         }
         debug!(target: "network", "dial to peer {:?} address {:?}", expected_peer_id, addr);
-        match Multihash::from_bytes(expected_peer_id.as_bytes().to_vec()) {
-            Ok(peer_id_hash) => {
-                addr.append(multiaddr::Protocol::P2p(peer_id_hash));
-                self.dial_addr(addr);
-            }
-            Err(err) => {
-                error!(target: "network", "failed to convert peer_id to addr: {}", err);
-            }
-        }
+        let p2p_phase =
+            Multihash::from_bytes(expected_peer_id.as_bytes().to_vec()).map_err(|_| {
+                Error::P2P(format!("peer_id is invalid {:?}", expected_peer_id).to_string())
+            })?;
+        addr.append(multiaddr::Protocol::P2p(p2p_phase));
+        self.p2p_control
+            .write()
+            .dial(addr, dial_protocol)
+            .map_err(|_| Error::P2P("dial error".to_string()))
+    }
+
+    pub fn dial(&self, expected_peer_id: &PeerId, addr: Multiaddr) -> Result<(), Error> {
+        self.inner_dial(expected_peer_id, addr, {
+            let mut protocols = vec![
+                PING_PROTOCOL_ID,
+                DISCOVERY_PROTOCOL_ID,
+                IDENTIFY_PROTOCOL_ID,
+            ];
+            protocols.extend(
+                // TODO remove the filter after refactor
+                self.ckb_protocols
+                    .iter()
+                    .map(|p| p.0.id())
+                    .filter(|id| *id != FEELER_PROTOCOL_ID)
+                    .collect::<Vec<_>>(),
+            );
+            DialProtocol::Multi(protocols)
+        })
```

Comment:
Have you tested in your machine? This dial will not work out. We have more protocols than those 4.

Reply:
This code dial more than 4 protocols on my machine.